lib/Analysis/ControlDependenceAnalysis.cpp

```diff
@@ -0,0 +1,333 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Declares some utility functions useful to analyzing the control dependencies
+// between basic blocks of the CFG. The control dependendies are calculated
+// using the algorithm from the following paper
+//   J.Ferrante, K.J. Ottenstein, and J. D. Warren, "The Program Dependence
+//   Graph and its Use in Optimizations", ACM Trans. Program. Lang. Syst., vol.
+//   9, pp. 319-349, 1987.
+//
+//===----------------------------------------------------------------------===//
+
+#include "dynamatic/Analysis/ControlDependenceAnalysis.h"
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+using namespace llvm;
+using namespace dynamatic;
+
+#define DEBUG_TYPE "cdg-analysis"
+
+void ControlDependenceAnalysis::identifyAllControlDeps(
+    mlir::func::FuncOp &funcOp) {
+  mlir::Region &funcReg = funcOp.getRegion();
+
+  // Initialize the all_control_deps_map by creating an entry for every block
+  // constituting the region
+  for (mlir::Block &block : funcReg.getBlocks()) {
+    llvm::SmallVector<mlir::Block *, 4> deps;
+    all_control_deps_map.insert(std::make_pair(&block, deps));
+  }
+
+  if (funcReg.hasOneBlock()) {
+    return; // functions made up of 1 basic block do not have any dependencies
+  }
+
+  // Get information about post-dominance
+  mlir::PostDominanceInfo postDomInfo;
+  // Get the post-dominance tree
+  llvm::DominatorTreeBase<mlir::Block, true> &postDomTree =
+      postDomInfo.getDomTree(&funcReg);
+
+  // Loop over the control flow edges connnecting the different blocks of this
+  // region
+  for (mlir::Block &block : funcReg.getBlocks()) {
+    for (mlir::Block *block_succ : block.getSuccessors()) {
+      if (!postDomInfo.properlyPostDominates(block_succ, &block)) {
+        mlir::Block *least_common_anc =
+            postDomInfo.findNearestCommonDominator(block_succ, &block);
+
+        if (least_common_anc == &block) {
+          // All nodes in the post-dominator tree on the path from the "block"
+          // to "block_succ" (including both "block" and "block_succ") should be
+          // control dependent on "block"
+
+          // easy case of block and block_succ
+          all_control_deps_map[&block].push_back(&block); // loop case
+          all_control_deps_map[block_succ].push_back(&block);
+
+          // traverse the tree to get all nodes between "least_common_anc" and
+          // "block_succ"
+          llvm::SmallVector<llvm::SmallVector<mlir::DominanceInfoNode *, 4>, 4>
+              traversed_nodes;
+          traversePostDomTree(&block, block_succ, &funcReg, &postDomTree,
+                              &traversed_nodes);
+
+          // prints the above traversal only in debug mode
+          printPostDomTreeTraversal(traversed_nodes);
```

Comment:
Correct me if I am wrong here: 

> You want to enumerate all possible paths from leastCommonAnc to block_succ and add all the nodes from all these paths to allControlDepsMap, 

is that correct?

---

If this is so, I think we can easily improve the readability of this part.

At the beginning of this file:
```c++
using PathInDomTree = SmallVector<DominanceInfoNode *, 4>;
```

And here, we could declare it as:

```
SmallVector<PathInDomTree, 4> allPathsFromLeastCommonAncToBlockSucc;
```

Then we could change the `traversePostDomTree` to `enumeratePathsInPostDomTree`